velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "velox/connectors/hive/storage_adapters/abfs/AbfsWriteFile.h"
+#include <azure/storage/files/datalake.hpp>
+
+namespace facebook::velox::filesystems::abfs {
+class BlobStorageFileClient final : public IBlobStorageFileClient {
+ public:
+  BlobStorageFileClient(const DataLakeFileClient client)
+      : client_(std::make_unique<DataLakeFileClient>(client)) {}
+
+  void create() override {
+    client_->Create();
+  }
+
+  PathProperties getProperties() override {
+    return client_->GetProperties().Value;
+  }
+
+  void append(const uint8_t* buffer, size_t size, uint64_t offset) override {
+    auto bodyStream = Azure::Core::IO::MemoryBodyStream(buffer, size);
+    client_->Append(bodyStream, offset);
+  }
+
+  void flush(uint64_t position) override {
+    client_->Flush(position);
+  }
+
+  void close() override {
+    // do nothing.
+  }
+
+ private:
+  std::unique_ptr<DataLakeFileClient> client_;
+};
+
+class AbfsWriteFile::Impl {
+ public:
+  explicit Impl(const std::string& path, const std::string& connectStr)
+      : path_(path), connectStr_(connectStr) {
+    // Make it a no-op if invoked twice.
+    if (position_ != -1) {
+      return;
+    }
+    position_ = 0;
+  }
+
+  void initialize() {
+    if (!blobStorageFileClient_) {
+      auto abfsAccount = AbfsAccount(path_);
+      auto fileClient = DataLakeFileClient::CreateFromConnectionString(
+          connectStr_, abfsAccount.fileSystem(), abfsAccount.filePath());
+      blobStorageFileClient_ = std::make_unique<BlobStorageFileClient>(
+          BlobStorageFileClient(fileClient));
+    }
+
+    VELOX_CHECK(!exist(), "File already exists");
+    blobStorageFileClient_->create();
+  }
+
+  /// mainly for test purpose.
+  void setFileClient(
+      std::shared_ptr<IBlobStorageFileClient> blobStorageManager) {
+    blobStorageFileClient_ = std::move(blobStorageManager);
+  }
+
+  bool exist() {
```

Comment:
This is only used once. Can we inline the body at the call site?

Reply:
hey @majetideepak   firstly, i super thank you for helping review. And recent weeks I am busying with internel project so didn't have change to resolve your comments. Hope you can understand,

And for this comment, you mean I can just move `blobStorageFileClient_ = std::move(blobStorageManager);` this part directly of part of body of `AbfsWriteFile::testingSetFileClient`  ?

Reply:
done!